### PR TITLE
Fix HID buffer swapping

### PR DESCRIPTION
**Pull request by ywwg**
HID Polling: only swap buffers on successful read

### CODE REVIEW

**Comment by uklotzde** [comment]
How often do we expect an unsuccessful read? If this occurs infrequently I suggest to zero out the buffer just before reading. Then only a single memset instruction would be required, reducing redundancy. Prior initialization of the buffer content becomes obsolete.

Initializing the buffer with a consistently is crucial here. Therefore repeating this code at multiple places is a code smell.

**Comment by uklotzde** [comment]
I just noticed that initializing a buffer with 0x00 or any other byte value is not sufficient. Reading a buffer [0x01] of length 1 is different than reading a buffer [0x01, 0x00] of length 2. But here those buffers would be considered equal. We need to take the actual length into account!

**Comment by uklotzde** [comment]
All buffers with trailing zeros would become ambiguous and their comparison could result in false positives.

**Review by daschuer — CHANGES_REQUESTED** [review]
Since we know the buffer length, zeroing out the buffer is a waste if time.
I think if we just store the buffer length as the fist byte, everything works.
It will fails reliable if the length is not equal, and memcmp automatically stops on the first fail.

File: src/controllers/hid/hidcontroller.cpp
Comment (Be-ing) [inline src/controllers/hid/hidcontroller.cpp:259]:
```suggestion
        const int nextBufferIndex = (m_iPollingBufferIndex + 1) % kNumBuffers;
```
Please spell out words

**Comment by Be-ing** [comment]
The buffer is passed directly as it is to the script, so I don't think manipulating the contents of the buffer is a great idea. That would get complicated, perhaps not be very performant, and isn't necessary. I think we can simply use the `bytesRead` return value from `hid_read` for the length of the buffer in the `memcmp` call. In case the current buffer is longer than the previous buffer, that is not a problem because the first byte (report ID) is different.

> Reading a buffer [0x01] of length 1 is different than reading a buffer [0x01, 0x00] of length 2.

This will not happen because every packet with the same first byte (report ID) is the same length.

**Comment by uklotzde** [comment]
> Since we know the buffer length, zeroing out the buffer is a waste if time.
> I think if we just store the buffer length as the fist byte, everything works.
> It will fails reliable if the length is not equal, and memcmp automatically stops on the first fail.

I agree. Storing the length as the first byte and only memcmp those number of bytes will produce correct results, regardless of how many bytes are contained in the previous buffer. No initialization needed, apart from setting the first byte of each buffer to 0.

**Comment by daschuer** [comment]
> The buffer is passed directly as it is to the script, so I don't think manipulating the contents of the buffer is a great idea

We need to pass it as &pCurrentBuffer[1] into the byte array to allow that the script can remain unchanged.

**Comment by Holzhaus** [comment]
Why not a second char array of length 2?

**Comment by uklotzde** [comment]
> Why not a second char array of length 2?

...because then we need to coordinate access between those two arrays. The in-place length prefix is convenient here.

**Comment by ywwg** [comment]
The length prefixing is "clever", but I agree with @Holzhaus that a second member variable is simpler and clearer.  We don't even need to track the size of both buffers, we only care if the current buffer is the same as the single previous buffer.  I don't have my controller to test at the moment but this seems legit.

File: src/controllers/hid/hidcontroller.cpp
Comment (uklotzde) [inline src/controllers/hid/hidcontroller.cpp:219]: Isn't this initialization obsolete now?

Comment (ywwg) [reply src/controllers/hid/hidcontroller.cpp:219]: it's a one-time memset, it seems worth initializing values to zero, esp because we pass these buffers to javascript. Better safe than sorry here.

**Comment by Be-ing** [comment]
The code looks good to me. @ywwg if you confirm this fixes the issue you were having with your controller, it's fine by me to merge it. But please test it with real hardware before merging.

File: src/controllers/hid/hidcontroller.cpp
Comment (uklotzde) [reply src/controllers/hid/hidcontroller.cpp:219]: I am not arguing about runtime overhead. This is completely redundant code and we never pass any uninitialized buffer to JavaScript. A comment in the code should clarify that. Otherwise, a future developer might wonder why this code is needed.

Comment (ywwg) [reply src/controllers/hid/hidcontroller.cpp:219]: if the read is <255 bytes, wouldn't the first version sent to javascript be bytesRead of data, and 255-bytesRead of junk?  I'll add a comment, but leave it in

Comment (uklotzde) [reply src/controllers/hid/hidcontroller.cpp:220]: The size of the array is always limited by the number of bytes read.

**Comment by uklotzde** [comment]
@ywwg Merging your own PR, moreover with pending approval by @daschuer is against our rules.

**Comment by ywwg** [comment]
oops, sorry.  I still have trouble with the "this is ok to merge but you can't do it yourself" model.  I can revert if needed

**Pull request merged**
